examples/relay-chat/index.js

```diff
@@ -0,0 +1,340 @@
+
+  const localPeerId = node.libp2p.peerId.toString();
+
+  const remotePeers = await node.libp2p.peerStore.all();
+  const remotePeerIds = new Set(remotePeers.map((peer) => peer.id.toString()));
+
+  ui.setRemotePeer(Array.from(remotePeerIds.keys()));
+
+  node.libp2p.addEventListener("peer:connect", async (event) => {
+    remotePeerIds.add(event.detail.toString());
+    ui.setRemotePeer(Array.from(remotePeerIds.keys()));
+    ui.setRelayMeshInfo(node.relay.gossipSub);
+  });
+
+  node.libp2p.addEventListener("peer:disconnect", (event) => {
+    remotePeerIds.delete(event.detail.toString());
+    ui.setRemotePeer(Array.from(remotePeerIds.keys()));
+    ui.setRelayMeshInfo(node.relay.gossipSub);
+  });
+
+  node.libp2p.addEventListener("peer:identify", (event) => {
+    const peer = event.detail;
+
+    if (!peer.protocols.includes("/webrtc-signaling/0.0.1")) {
+      return;
+    }
+
+    ui.setWebrtcPeer(peer.peerId.toString());
+    ui.setRelayMeshInfo(node.relay.gossipSub);
+  });
+
+  window.node = node;
+
+  return {
+    unsubscribeFromMessages,
+    info: {
+      contentTopic,
+      localPeerId,
+    },
+    sendMessage: async ({ text, nick }) => {
+      if (!text || !nick) {
+        return;
+      }
+
+      const protoMessage = ChatMessage.create({
+        nick,
+        timestamp: Date.now(),
+        text: utf8ToBytes(text),
+      });
+
+      await node.relay.send(Encoder, {
+        payload: ChatMessage.encode(protoMessage).finish(),
+      });
+    },
+    dial: async (multiaddr) => {
+      ui.setStatus("connecting...", "progress");
+      await node.dial(multiaddr);
+      ui.setStatus("connected", "success");
+    },
+    dialWebRTCpeer: async (peerId) => {
+      const peers = await node.libp2p.peerStore.all();
+      const circuitPeer = peers.filter(
+        (p) =>
+          p.protocols.includes("/libp2p/circuit/relay/0.2.0/hop") &&
+          p.protocols.includes("/libp2p/circuit/relay/0.2.0/stop")
+      )[0];
+
+      if (!circuitPeer) {
+        throw Error("No Circuit peer is found");
+      }
+
+      let multiaddr = circuitPeer.addresses.pop().multiaddr;
+      multiaddr = `${multiaddr}/p2p/${circuitPeer.id.toString()}/p2p-circuit/webrtc/p2p/${peerId}`;
+
+      await node.dial(multiaddr);
+      ui.setRelayMeshInfo(node.relay.gossipSub);
+    },
+    ensureWebRTCconnectionInRelayMesh: async () => {
+      const promises = node.libp2p
+        .getConnections()
+        .filter((c) => c.stat.multiplexer === "/webrtc")
+        .map(async (c) => {
+          const outboundStream = node.relay.gossipSub.streamsOutbound.get(
+            c.remotePeer.toString()
+          );
+          const isWebRTCOutbound =
+            outboundStream.rawStream.constructor.name === "WebRTCStream";
+
+          if (isWebRTCOutbound) {
+            return;
+          }
+
+          node.relay.gossipSub.streamsOutbound.delete(c.remotePeer.toString());
+          await node.relay.gossipSub.createOutboundStream(
```

Comment:
This is the main problem with the current approach, due to how `WebRTC` transport is made it is able only to create `outbound` connection using `WebRTCStream` but `inbound` stays `MplexStream` which is wrong, and prevents the mesh from functioning over `WebRTC` transport. 

Anyway we don't intend to limit ourselves to the `libp2p`'s WebRTC transport and we need a different solution on which I am working: it is to establish `WebRTC` connection similarly to how we do it in [noise-rtc](https://github.com/waku-org/js-waku-examples/tree/master/examples/noise-rtc) by using `contentTopics`. Main thing that makes everything ugly here is to manually put streams and connections into `gossipSub`

Reply:
> but inbound stays MplexStream which is wrong, and prevents the mesh from functioning over WebRTC transport.

Is that a libp2p bug?

Reply:
I would say it is `gossipSub`'s bug since it is failing to upgrade `inbound` connection from mplex to webrtc due to the condition explicitly skipping it and it happens to be that webrtc transport starts as mplex and then get's upgraded 

I am looking at how to fix it in `js-libp2p-gossipsub` but I consider it to be a low-priority